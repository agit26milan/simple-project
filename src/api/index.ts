import axios from 'axios';

const baseURL = 'https://auntieanne-demo.proseller.io/product/api';

export const getListMenu = async (id: string) => {
  const url = `/productpreset/loadcategory/webOrdering/${id}`;
  return await axios({method: 'get', baseURL, url});
};

export const getSublistMenu = async (companyId: string, id: string) => {
  const url = `/productpreset/loaditems/webOrdering/${companyId}/${id}`;
  return await axios({method: 'post', baseURL, url});
};
