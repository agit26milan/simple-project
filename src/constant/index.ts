export const COMPANY_ID = '64e4e47b-b428-4ee3-97f5-62bfa03c8ba6';
