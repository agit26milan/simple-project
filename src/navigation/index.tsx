import * as React from 'react';
import {createStackNavigator} from '@react-navigation/stack';
import Tab from './TopTab';
import FavoriteScreen from '../screen/Favorite';

const Stack = createStackNavigator();

const RootNavigation = () => {
  return (
    <Stack.Navigator>
      <Stack.Screen
        options={{headerShown: false}}
        name="BottomTab"
        component={Tab}
      />
      <Stack.Screen name="Favorite" component={FavoriteScreen} />
    </Stack.Navigator>
  );
};

export default RootNavigation;
