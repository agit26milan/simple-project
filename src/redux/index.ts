import {configureStore} from '@reduxjs/toolkit';
import {useDispatch, useSelector, TypedUseSelectorHook} from 'react-redux';
import rootReducer from './reducer';
import {persistStore, persistReducer} from 'redux-persist';
import AsyncStorage from '@react-native-async-storage/async-storage';

const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
  whitelist: ['favorite'],
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      serializableCheck: false,
    }),
  devTools: __DEV__,
});

export const persistor = persistStore(store);
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
export type AppStore = typeof store;
// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
