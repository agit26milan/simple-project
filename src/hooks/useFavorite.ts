import {useDispatch, useSelector} from 'react-redux';
import {saveToFavorite} from '../redux/action/favorite/action';
import {RootState} from '../redux';

const useFavorite = () => {
  const listFav = useSelector((state: RootState) => state.favorite.data);
  const dispatch = useDispatch();
  const onSaveFavorite = (selectProduct: Home.ResponseSubMenu) => {
    if (selectProduct) {
      dispatch(saveToFavorite(selectProduct));
    }
  };

  const checkIsOnFavorite = (id?: string) => {
    const findINdex = listFav.findIndex(
      (data: Home.ResponseSubMenu) => data.product.id === id,
    );
    return findINdex > -1;
  };

  return {
    onSaveFavorite,
    checkIsOnFavorite,
  };
};

export default useFavorite;
