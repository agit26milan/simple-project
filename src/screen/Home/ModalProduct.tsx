import React from 'react';
import ModalGlobal from '../../components/ModalGlobal';
import {View, Image, Text, StyleSheet, TouchableOpacity} from 'react-native';
import {formatMoney} from '../../utils';
import useFavorite from '../../hooks/useFavorite';

const ModalProduct: React.FC<Modal.ModalProduct> = props => {
  const {selectProduct} = props;
  const {checkIsOnFavorite} = useFavorite();
  const onSave = () => {
    if (props.onSaveFavoriteHandle && selectProduct) {
      props.onSaveFavoriteHandle(selectProduct);
    }
  };

  const onRemove = () => {
    if (props.onRemoveFavorite && selectProduct) {
      props.onRemoveFavorite(selectProduct);
    }
  };

  return (
    <ModalGlobal {...props}>
      <View style={styles.containerModal}>
        <Image
          style={{height: 400}}
          source={{uri: selectProduct?.product.defaultImageURL}}
          testID="image"
        />
        <View style={styles.mainContainer}>
          <Text testID="productName" style={styles.productNameText}>
            {selectProduct?.product.name}
          </Text>
          <Text testID="description" style={styles.descriptionText}>
            {selectProduct?.product.description}
          </Text>
          <Text testID="price" style={styles.priceText}>
            $ {formatMoney(String(selectProduct?.product.retailPrice))}{' '}
          </Text>
        </View>

        <View style={{position: 'absolute', bottom: 0, width: '100%'}}>
          {checkIsOnFavorite(selectProduct?.product.id) ? (
            <TouchableOpacity
              onPress={onRemove}
              style={styles.unfavoriteButton}>
              <Text style={{color: 'white'}}>Remove favorite</Text>
            </TouchableOpacity>
          ) : (
            <TouchableOpacity onPress={onSave} style={styles.favoriteButton}>
              <Text style={{color: 'white'}}>Add to favorite</Text>
            </TouchableOpacity>
          )}
        </View>
      </View>
    </ModalGlobal>
  );
};

const styles = StyleSheet.create({
  containerModal: {
    flex: 1,
    backgroundColor: 'white',
  },
  mainContainer: {
    paddingHorizontal: 10,
    marginTop: 20,
  },
  productNameText: {
    fontSize: 16,
    fontWeight: 'bold',
    marginBottom: 10,
  },
  descriptionText: {
    fontSize: 14,
    marginBottom: 10,
  },
  priceText: {
    fontSize: 16,
    fontWeight: 'bold',
    marginBottom: 10,
  },
  favoriteButton: {
    backgroundColor: '#06c',
    marginTop: 20,
    alignItems: 'center',
    justifyContent: 'center',
    padding: 10,
  },
  unfavoriteButton: {
    backgroundColor: 'red',
    marginTop: 20,
    alignItems: 'center',
    justifyContent: 'center',
    padding: 10,
  },
});

export default ModalProduct;
