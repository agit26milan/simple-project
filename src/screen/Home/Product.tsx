import React from 'react';
import {
  Text,
  TouchableOpacity,
  View,
  StyleSheet,
  Image,
  Dimensions,
} from 'react-native';
import {formatMoney} from '../../utils';
import Love from '../../assets/images/favourite.png';
import useFavorite from '../../hooks/useFavorite';
export type PropsDataModifier = {
  item: Home.ResponseSubMenu;
  numColumn: number;
  onPress: (item: Home.ResponseSubMenu) => void;
};

const Product: React.FC<PropsDataModifier> = props => {
  const {item, numColumn} = props;
  const widthProduct = Dimensions.get('window').width / (numColumn || 3) - 10;
  const {checkIsOnFavorite} = useFavorite();
  return (
    <TouchableOpacity
      onPress={() => props.onPress(item)}
      style={styles.containerButton}
      testID="buttonPress">
      <View style={{width: widthProduct}}>
        <Image
          style={{
            height: widthProduct,
            width: widthProduct,
            borderRadius: widthProduct / 2,
          }}
          source={{uri: item.product.defaultImageURL}}
          resizeMode="cover"
          testID="image"
        />
        <Text testID="product-name" style={styles.titleText}>
          {item.product.name}
        </Text>
        <Text testID="money" style={styles.priceText}>
          $ {formatMoney(String(item.product.retailPrice))}{' '}
        </Text>
        {checkIsOnFavorite(item.product.id) && (
          <View testID="showFavButton" style={{alignItems: 'flex-end'}}>
            <Image style={{height: 24, width: 24}} source={Love} />
          </View>
        )}
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  containerButton: {
    marginHorizontal: 5,
    marginBottom: 15,
  },
  titleText: {
    marginTop: 5,
  },
  priceText: {
    fontSize: 16,
    fontWeight: 'bold',
  },
});

export default React.memo(Product);
