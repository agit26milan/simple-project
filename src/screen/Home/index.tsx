import React from 'react';
import {
  View,
  StyleSheet,
  FlatList,
  RefreshControl,
  Text,
  TouchableOpacity,
  Image,
} from 'react-native';
import {getSublistMenu} from '../../api';
import {COMPANY_ID} from '../../constant';
import Product from './Product';
import ModalProduct from './ModalProduct';
import useFavorite from '../../hooks/useFavorite';
import {useNavigation} from '@react-navigation/native';
import Love from '../../assets/images/favourite.png';

const HomeScreen = (props: any) => {
  const {route} = props;
  const [listSubMenu, setListSubMenu] = React.useState<Home.ResponseSubMenu[]>(
    [],
  );
  const {navigate} = useNavigation();
  const {onSaveFavorite} = useFavorite();
  const [openModal, setOpenModal] = React.useState<boolean>(false);
  const [loading, setLoading] = React.useState<boolean>(false);
  const [selectProduct, setSelectProduct] =
    React.useState<Home.ResponseSubMenu | null>(null);
  const getSubListCategory = async () => {
    try {
      setLoading(true);
      const response = await getSublistMenu(COMPANY_ID, route.params.id);
      const {data} = response;
      setListSubMenu(data.data);
      setLoading(false);
    } catch (e) {
      setLoading(false);
    }
  };

  const onPullRefresh = () => {
    getSubListCategory();
  };

  const onClickProduct = (product: Home.ResponseSubMenu) => {
    setOpenModal(true);
    setSelectProduct(product);
  };

  const closeModal = () => {
    setOpenModal(false);
  };

  React.useEffect(() => {
    if (route.params?.id) {
      getSubListCategory();
    }
  }, [route.params]);

  const onSaveFavoriteHandle = () => {
    if (selectProduct) {
      onSaveFavorite(selectProduct);
    }
  };

  const renderItem = ({item}: any) => (
    <Product item={item} numColumn={3} onPress={onClickProduct} />
  );

  const goToFavoritePage = () => {
    navigate('Favorite');
  };

  return (
    <View style={styles.containerMenu}>
      <FlatList
        refreshControl={
          <RefreshControl refreshing={loading} onRefresh={onPullRefresh} />
        }
        data={listSubMenu}
        renderItem={renderItem}
        numColumns={3}
      />
      <ModalProduct
        onBackButtonPress={closeModal}
        onBackdropPress={closeModal}
        isVisible={openModal}
        useNativeDriver={true}
        selectProduct={selectProduct}
        onSaveFavoriteHandle={onSaveFavoriteHandle}
      />
      <View>
        <TouchableOpacity style={styles.floatButton} onPress={goToFavoritePage}>
          <Image source={Love} />
        </TouchableOpacity>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  containerMenu: {
    paddingTop: 20,
    backgroundColor: 'white',
    flex: 1,
  },
  titleText: {
    fontSize: 18,
    fontWeight: 'bold',
  },
  containerModal: {
    flex: 1,
    backgroundColor: 'white',
  },
  mainContainer: {
    paddingHorizontal: 10,
  },
  productNameText: {
    fontSize: 16,
    fontWeight: 'bold',
    marginBottom: 10,
  },
  descriptionText: {
    fontSize: 14,
    marginBottom: 10,
  },
  priceText: {
    fontSize: 16,
    fontWeight: 'bold',
    marginBottom: 10,
  },
  favoriteButton: {
    backgroundColor: '#06c',
    marginTop: 20,
    alignItems: 'center',
    justifyContent: 'center',
    padding: 10,
  },
  floatButton: {
    position: 'absolute',
    bottom: 50,
    right: 20,
    backgroundColor: 'transparent',
    height: 50,
    borderRadius: 25,
  },
});

export default React.memo(HomeScreen);
