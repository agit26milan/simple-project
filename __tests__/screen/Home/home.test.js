import React from 'react';
import {fireEvent, render, screen} from '@testing-library/react-native';
import Home from '../../../src/screen/Home';
import ReduxProvider from '../../../src/reduxTest';
import {store} from '../../../src/redux';
import Product from '../../../src/screen/Home/Product';
import useFavorite from '../../../src/hooks/useFavorite';
import ModalProduct from '../../../src/screen/Home/ModalProduct';

const mockApi = {
  name: 'Seaweed',
  sequence: 1,
  itemType: 'PRODUCT',
  id: '1645669665888',
  productID: 'product::1305bdc4-f9f7-4fc0-b6cc-7da7aeec863a',
  product: {
    id: '1305bdc4-f9f7-4fc0-b6cc-7da7aeec863a',
    name: 'Seaweed',
    description: 'Seaweed nori, lightly salted. A delicious healthy option.',
    retailPrice: 10000000,
    categoryName: 'PRETZELS',
    defaultImageURL:
      'https://cdn-bucket-file-manager.s3.ap-southeast-1.amazonaws.com/Upload/e902cee7-6650-447e-989a-9209944192fc/product/e79e0ffd-d0e8-4a69-9080-b78526de3d91.jpg',
    orderingStatus: 'AVAILABLE',
    code: 'PRD0000012',
    promotions: [],
    orderingAvaibility: {
      storePickUp: true,
      delivery: true,
      storeCheckOut: true,
      takeAway: true,
      dineIn: true,
    },
  },
};

jest.mock('redux-persist', () => {
  const real = jest.requireActual('redux-persist');
  return {
    ...real,
    persistReducer: jest
      .fn()
      .mockImplementation((config, reducers) => reducers),
  };
});

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({
    navigate: jest.fn(),
  }),
  useRoute: () => ({
    params: {
      id: '123',
    },
  }),
}));

jest.mock('../../../src/api');

describe('it should be match with snapshot', () => {
  const wrapper = ({children}) => (
    <ReduxProvider reduxStore={store}>{children}</ReduxProvider>
  );
  it('Home Code should be not change', () => {
    const {toJSON} = render(<Home route={{params: '123'}} />, {wrapper});
    expect(toJSON).toMatchSnapshot();
  });

  it('product list should correct', async () => {
    const mockPress = jest.fn();
    const {getByTestId} = render(
      <Product item={mockApi} onPress={mockPress} />,
      {wrapper},
    );
    await fireEvent.press(getByTestId('buttonPress'));
    expect(mockPress).toHaveBeenCalledTimes(1);
    expect(getByTestId('image').props.source.uri).toEqual(
      mockApi.product.defaultImageURL,
    );
    expect(getByTestId('product-name').props.children).toEqual(
      mockApi.product.name,
    );
    expect(getByTestId('money').props.children).toEqual([
      '$ ',
      '10.000.000',
      ' ',
    ]);
  });

  it('Modal Product should correct', async () => {
    const {getByTestId} = render(
      <ModalProduct selectProduct={mockApi} isVisible={true} />,
      {wrapper},
    );
    expect(getByTestId('image').props.source.uri).toEqual(
      mockApi.product.defaultImageURL,
    );
    expect(getByTestId('productName').props.children).toEqual(
      mockApi.product.name,
    );
    expect(getByTestId('description').props.children).toEqual(mockApi.product.description);
    expect(getByTestId('price').props.children).toEqual([
      '$ ',
      '10.000.000',
      ' ',
    ]);
  });
});