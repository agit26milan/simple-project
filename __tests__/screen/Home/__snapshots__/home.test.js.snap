// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should be match with snapshot Home Code should be not change 1`] = `[Function]`;
